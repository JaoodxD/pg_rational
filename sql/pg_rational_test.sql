create extension pg_rational;

-- input

-- can parse a simple fraction
select '1/3'::rational;
-- can parse negatives
select '-1/3'::rational;
select '1/-3'::rational;

-- SEND works
select rational_send('1/3');

-- too big
select '9223372036854775808/9223372036854775807'::rational;
-- no spaces
select '1 /3'::rational;
-- no zero denominator
select '1/0'::rational;
-- no single numbers
select '1'::rational;
-- no garbage
select ''::rational;
select 'sdfkjsdfj34984538'::rational;

-- simplification

-- double negative becomes positive
select rational_simplify('-1/-3');
-- works with negative value
select rational_simplify('-3/12');
-- don't move negative if it would overflow
select rational_simplify('1/-9223372036854775808');
-- biggest value reduces
select rational_simplify('9223372036854775807/9223372036854775807');
-- smallest value reduces
select rational_simplify('-9223372036854775808/-9223372036854775808');
-- idempotent on simplified expression
select rational_simplify('1/1');

-- addition

-- additive identity
select '0/1'::rational + '1/2';
-- additive inverse
select '1/2'::rational + '-1/2';
-- just regular
select '1/2'::rational + '1/2';
-- forcing intermediate simplification
select '9223372036854775807/9223372036854775807'::rational + '1/1';
-- overflow (sqrt(max)+1)/1 + 1/sqrt(max)
select '3037000501/1'::rational + '1/3037000500';
