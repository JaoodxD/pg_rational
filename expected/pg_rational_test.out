create extension pg_rational;
-- input
-- can parse a simple fraction
select '1/3'::rational;
 rational 
----------
 1/3
(1 row)

-- can parse negatives
select '-1/3'::rational;
 rational 
----------
 -1/3
(1 row)

select '1/-3'::rational;
 rational 
----------
 -1/3
(1 row)

-- SEND works
select rational_send('1/3');
           rational_send            
------------------------------------
 \x00000000000000010000000000000003
(1 row)

-- too big
select '9223372036854775808/9223372036854775807'::rational;
ERROR:  numerator or denominator outside valid int64 value: "9223372036854775808/9223372036854775807"
LINE 1: select '9223372036854775808/9223372036854775807'::rational;
               ^
-- no spaces
select '1 /3'::rational;
ERROR:  invalid input syntax for fraction: "1 /3"
LINE 1: select '1 /3'::rational;
               ^
-- no zero denominator
select '1/0'::rational;
ERROR:  fraction cannot have zero denominator: "1/0"
LINE 1: select '1/0'::rational;
               ^
-- no single numbers
select '1'::rational;
ERROR:  invalid input syntax for fraction: "1"
LINE 1: select '1'::rational;
               ^
-- no garbage
select ''::rational;
ERROR:  invalid input syntax for fraction: ""
LINE 1: select ''::rational;
               ^
select 'sdfkjsdfj34984538'::rational;
ERROR:  invalid input syntax for fraction: "sdfkjsdfj34984538"
LINE 1: select 'sdfkjsdfj34984538'::rational;
               ^
-- simplification
-- double negative becomes positive
select rational_simplify('-1/-3');
 rational_simplify 
-------------------
 1/3
(1 row)

-- works with negative value
select rational_simplify('-3/12');
 rational_simplify 
-------------------
 -1/4
(1 row)

-- don't move negative if it would overflow
select rational_simplify('1/-9223372036854775808');
   rational_simplify    
------------------------
 1/-9223372036854775808
(1 row)

-- biggest value reduces
select rational_simplify('9223372036854775807/9223372036854775807');
 rational_simplify 
-------------------
 1/1
(1 row)

-- smallest value reduces
select rational_simplify('-9223372036854775808/-9223372036854775808');
 rational_simplify 
-------------------
 1/1
(1 row)

-- idempotent on simplified expression
select rational_simplify('1/1');
 rational_simplify 
-------------------
 1/1
(1 row)

-- addition
-- additive identity
select '0/1'::rational + '1/2';
 ?column? 
----------
 1/2
(1 row)

-- additive inverse
select '1/2'::rational + '-1/2';
 ?column? 
----------
 0/4
(1 row)

-- just regular
select '1/2'::rational + '1/2';
 ?column? 
----------
 4/4
(1 row)

-- forcing intermediate simplification
select '9223372036854775807/9223372036854775807'::rational + '1/1';
 ?column? 
----------
 2/1
(1 row)

-- overflow (sqrt(max)+1)/1 + 1/sqrt(max)
select '3037000501/1'::rational + '1/3037000500';
ERROR:  intermediate value overflow in rational addition
